specs/builder.go

```diff
@@ -0,0 +1,256 @@
+package specs
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/moov-io/iso8583"
+	"github.com/moov-io/iso8583/encoding"
+	"github.com/moov-io/iso8583/field"
+	"github.com/moov-io/iso8583/padding"
+	"github.com/moov-io/iso8583/prefix"
+)
+
+const (
+	FieldPrefix = "Field"
+)
+
+var Builder MessageSpecBuilder = &messageSpecBuilder{}
+
+type MessageSpecBuilder interface {
+	ImportJSON([]byte) (*iso8583.MessageSpec, error)
+	ExportJSON(spec *iso8583.MessageSpec) ([]byte, error)
+}
+
+type messageSpecBuilder struct{}
+
+type specDummy struct {
+	Name   string                `json:"name,omitempty" xml:"name,omitempty"`
+	Fields map[string]fieldDummy `json:"fields,omitempty" xml:"fields,omitempty"`
+}
+
+type fieldDummy struct {
+	Type         string    `json:"type,omitempty" xml:"type,omitempty"`
+	Length       int       `json:"length,omitempty" xml:"length,omitempty"`
+	IDLength     int       `json:"id_length,omitempty" xml:"id_length,omitempty"`
+	Description  string    `json:"description,omitempty" xml:"description,omitempty"`
+	Enc          string    `json:"enc,omitempty" xml:"enc,omitempty"`
+	Prefix       string    `json:"prefix,omitempty" xml:"prefix,omitempty"`
+	PrefixLength int       `json:"prefix_length,omitempty" xml:"prefix_length,omitempty"`
+	Padding      *padDummy `json:"padding,omitempty" xml:"padding,omitempty"`
+}
+
+type padDummy struct {
+	Type string `json:"type" xml:"type"`
+	Pad  rune   `json:"pad" xml:"pad"`
+}
+
+func (builder *messageSpecBuilder) ImportJSON(raw []byte) (*iso8583.MessageSpec, error) {
+	dummy := specDummy{}
+	err := json.Unmarshal(raw, &dummy)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(dummy.Fields) == 0 {
+		return nil, fmt.Errorf("invalid json spec file")
+	}
+
+	spec := iso8583.MessageSpec{
+		Name:   dummy.Name,
+		Fields: make(map[int]field.Field),
+	}
+
+	for key, dummyField := range dummy.Fields {
+		index := 0
+		_, err := fmt.Sscanf(key, FieldPrefix+"%d", &index)
+		if err != nil {
+			return nil, fmt.Errorf("invalid field index, index's format is `Field`+index ")
+		}
+
+		enc := getEncodingInterface(dummyField.Enc)
+		pref := getPrefixInterface(dummyField.Prefix, dummyField.PrefixLength)
+		pad := getPadInterface(dummyField.Padding)
+
+		var newField field.Field
+
+		switch dummyField.Type {
+		case reflect.TypeOf(field.String{}).Name():
+			newField = field.NewString(&field.Spec{
```

Comment:
I suggest here two things:
1. let's define spec before `switch` like this

```go
spec: = &field.Spec{
		Length:      dummyField.Length,
		IDLength:    dummyField.IDLength,
		Description: dummyField.Description,
		Enc:         enc,
		Pref:        pref,
		Pad:         pad,
	}
```

and then, let's create a map for the field constructor, like this:

```
var fieldsConstructor = map[string]func(spec *field.Spec) *field.Field {
  "String": func(spec *field.Spec) { return field.NewString(spec) },
  "Numeric": func(spec *field.Spec) { return field.NewNumeric(spec) },
  // ...
}
```

then inside we can do:

```go
fieldConstructor := fieldsConstructor[dummyField.Type]
if fieldConstructor == nil {
  // handle error
}

newField := fieldConstructor(spec)

// ...
```

Reply:
fixed